File: frontend/src/types/report.ts
export interface RevenueOut {
  total_current: number
  total_previous: number
  variation_pct: number | null
}

export interface TicketOut {
  average_current: number | null
  average_previous: number | null
  variation_pct: number | null
}

export interface VolumeOut {
  orders_current: number
  orders_previous: number
  variation_pct: number | null
}

export interface UnitsOut {
  units_current: number
  units_previous: number
  variation_pct: number | null
}

export interface NewCustomersOut {
  new_customers_count: number
  returning_customers_count: number
  new_customers_revenue: number
  returning_customers_revenue: number
}

export interface ProductMarginOut {
  product_key: string
  product_name: string
  revenue: number
  cost: number
  margin_value: number
  margin_pct: number | null
  cost_available: boolean
}

export interface AggregateMarginOut {
  revenue_with_cost_known: number
  cost_total: number
  margin_value: number
  margin_pct: number | null
  revenue_coverage_pct: number
}

export interface RankingEntryOut {
  name: string
  revenue_current: number
  revenue_previous: number
  variation_pct: number | null
  variation_abs: number
}

export interface BestsellerOut {
  product_key: string
  product_name: string
  units_sold: number
  revenue: number
}

export interface HighestPricedSaleOut {
  product_key: string
  product_name: string
  unit_price: number
  order_id: string
}

export interface ChurnedCustomerOut {
  customer_id: string
  last_purchase_at: string
  days_since_last_purchase: number
  avg_interval_days: number
}

export interface InsightOut {
  key: string
  category: string
  title: string
  description: string
  score: number
}

export interface ReportOut {
  period_start: string
  period_end: string
  revenue: RevenueOut
  ticket: TicketOut
  volume: VolumeOut
  units: UnitsOut
  new_customers: NewCustomersOut
  product_margins: ProductMarginOut[]
  aggregate_margin: AggregateMarginOut
  negative_margin_products: ProductMarginOut[]
  ranking_products: RankingEntryOut[]
  ranking_categories: RankingEntryOut[]
  churned_customers: ChurnedCustomerOut[]
  bestseller: BestsellerOut | null
  highest_priced_sale: HighestPricedSaleOut | null
  insights: InsightOut[]
  last_synced_at: string | null
  last_sync_label: string | null
  validation_warnings: string[]
}

export interface DataSourceOut {
  source_type: string
  last_synced_at: string | null
  last_sync_label: string | null
  last_validation_warnings: string[]
}

export interface IngestionResultOut {
  orders_created: number
  orders_updated: number
  items_created: number
  warnings: string[]
  row_errors: string[]
}

File: backend/app/schemas/report.py
from datetime import date, datetime

from pydantic import BaseModel


class RevenueOut(BaseModel):
    total_current: float
    total_previous: float
    variation_pct: float | None


class TicketOut(BaseModel):
    average_current: float | None
    average_previous: float | None
    variation_pct: float | None


class VolumeOut(BaseModel):
    orders_current: int
    orders_previous: int
    variation_pct: float | None


class UnitsOut(BaseModel):
    units_current: float
    units_previous: float
    variation_pct: float | None


class NewCustomersOut(BaseModel):
    new_customers_count: int
    returning_customers_count: int
    new_customers_revenue: float
    returning_customers_revenue: float


class ProductMarginOut(BaseModel):
    product_key: str
    product_name: str
    revenue: float
    cost: float
    margin_value: float
    margin_pct: float | None
    cost_available: bool


class AggregateMarginOut(BaseModel):
    revenue_with_cost_known: float
    cost_total: float
    margin_value: float
    margin_pct: float | None
    revenue_coverage_pct: float


class RankingEntryOut(BaseModel):
    name: str
    revenue_current: float
    revenue_previous: float
    variation_pct: float | None
    variation_abs: float


class ChurnedCustomerOut(BaseModel):
    customer_id: str
    last_purchase_at: date
    days_since_last_purchase: int
    avg_interval_days: float


class InsightOut(BaseModel):
    key: str
    category: str
    title: str
    description: str
    score: float


class ReportOut(BaseModel):
    period_start: date
    period_end: date
    revenue: RevenueOut
    ticket: TicketOut
    volume: VolumeOut
    units: UnitsOut
    new_customers: NewCustomersOut
    product_margins: list[ProductMarginOut]
    aggregate_margin: AggregateMarginOut
    negative_margin_products: list[ProductMarginOut]
    ranking_products: list[RankingEntryOut]
    ranking_categories: list[RankingEntryOut]
    churned_customers: list[ChurnedCustomerOut]
    insights: list[InsightOut]
    last_synced_at: datetime | None
    last_sync_label: str | None
    validation_warnings: list[str]
